language/src/ceylon/language/Sequence.ceylon

```diff
@@ -287,8 +287,13 @@ shared sealed interface Sequence<out Element=Anything>
         
         first => outer.first;
         size => outer.size*times;
-        rest => sublistFrom(1).sequence(); //TODO!
-        
+        rest => if (times==1)
+                    then outer.rest
+                    else let (tail = Repeat(times-1))
+                         if (outer.longerThan(1))
+                             then outer.rest.append(tail)
+                             else tail;
+
         getFromFirst(Integer index)
```

Comment:
> * The JVM backend incorrectly interprets the original `sublistFrom(1)` as `outer.sublistFrom(1)`, but it should be `this.sublistFrom(1)`.

I agree that it should be interpreted as a call to `Rest.sublistFrom(1)`. If it's not, then that's a bug in the backend and we should open a separate issue.

> * The Dart & JS backends incorrectly interpret `Repeat(times-1)` as `this.Repeat(times-1)`, but it should be `outer.Repeat(times-1)`. `Repeat` doesn't have a `Repeat` member, since `Sequence.Repeat` is not `shared`, and, the typechecker disallows `this.Repeat(...)`. The counter point to the second item is that since `Sequence.Repeat` is visible inside `Repeat`, the following *is* allowed inside `Repeat`: `(this of Repeat).Repeat(...)`. Can the implicit `this` be implicitly widened to `Sequence` to make it a suitable receiver for `Repeat`!?...

My reading of the spec inclines me toward the *second* interpretation.

1. `Sequence.Repeat` is visible at the program element, since it is in scope. 

2. Furthermore, the spec says unambiguously that:
    
    > a declaration inherited by a nested class or interface hides a declaration of the containing body

I guess it's a bug that the expression `this.Repeat` is disallowed by the typechecker. That's inconsistent. The spec says the following, which does not precisely address this situation, but seems to imply that it would be visible:

> if the qualifying type inherits a class or interface that contains the program element, and an un-shared declaration contained directly in the body of this class or interface has the same name as a shared member of the qualifying type, in which case the un-shared declaration hides the shared member

In this case _the qualifying type_ would be `Rest`, the _inherit[ed] class or interface that contains the program element_ would be `Sequence`, and the _un-shared declaration contained directly in the body of this class or interface_ would be `Rest`. It seems clear that the spec expects such un-shared declarations to be visible.

I will report that as a minor bug against the typechecker.